examples/assembly/addarrays.s

```diff
@@ -0,0 +1,116 @@
+.globl addarrays_serial
+.type addarrays_serial, %function
+
+# void addarrays_serial(int *res, int *A, int *B, int N)
+# Set res[i] = A[i] + B[i] for i from [0, N)
+# x0 -> res
+# x1 -> A
+# x2 -> B
+# w3 -> N
+# w4 -> loop counter
+
+addarrays_serial:
+    mov w4, wzr
+.loop_begin:
+    cmp w4, w3
+    beq .loop_exit
+    ldr w5, [x1, w4, uxtw 2]
+    ldr w6, [x2, w4, uxtw 2]
+    add w5, w5, w6
+    str w5, [x0, w4, uxtw 2]
+    add w4, w4, 1
+    b .loop_begin
+.loop_exit:
+    ret
+
+.global addarrays_sve
+.type addarrays_sve, %function
+
+# void addarrays_sve (int *restrict res, int *A, int *B, int N)
+# Set res[i] = A[i] + B[i] for i from [0, N) using SVE vectors.
+# x0 -> res
+# x1 -> A
+# x2 -> B
+# x3 -> N
+# x4 -> loop counter
+
+addarrays_sve:
+    mov x4, xzr
+.loop_begin_2:
+    whilelt p0.s, x4, x3
+    b.none .loop_exit_2
+    ld1w z0.s, p0/z, [x1, x4, lsl 2]
+    ld1w z1.s, p0/z, [x2, x4, lsl 2]
+    add z0.s, z0.s, z1.s
+    st1w z0.s, p0, [x0, x4, lsl 2]
+    incw x4
+    b .loop_begin_2
+.loop_exit_2:
+    ret
+
+.global addarrays_neon
+.type addarrays_neon, %function
+
+addarrays_neon:
+    mov w4, 0
+    lsr w5, w3, 2
+.vector_loop:
+    cmp w4, w5
+    beq .vector_loop_exit
+    ldr q30, [x1, x4, lsl 4]
+    ldr q31, [x2, x4, lsl 4]
+    add v30.4s, v30.4s, v31.4s
+    str q30, [x0, x4, lsl 4]
+    add w4, w4, 1
+    b .vector_loop
+.vector_loop_exit:
+    lsl w4, w5, 2
```

Comment:
This bit I don't understand. Above you have:
```
lsr w5, w3, 2
```
So you're dividing the loop count by 4, because 4 integers fit into 128 bits.

Now you're recovering that loop count by multiplying it by 4? Surely if my loop count was for example, 6, shifting down 2 loses the remainder of 2. We're not going to get that back by shifting it back up 2.

But maybe I have the wrong end of the stick here.

Reply:
w5 contains the number of iteratios for vectorized loop and for scalar tail loop we set loop counter to w5 * 4 so it iterates over the remaining iterations. For n = 6, the number of vectorized iterations will be 6 >> 2 = 1, and the tail loop will iterate from [4, 6).

Equivalent C code:
int vec_niters = n / 4;
for (int i = 0; i < vec_niters; i++)
  {
    int32x4_t vA = vld1q_s32 (&a[i * 4]);
    int32x4_t vB = vld1q_s32 (&b[i * 4]);
    vA += vB;
    vst1q_s32 (&res[i * 4], vA);
  }

for (int i = vec_niters * 4; i < n; i++)
  res[i] = a[i] + b[i];

Reply:
Alternatively, I guess we could do:
```
int vec_upper_bound = (n / 4) * 4;
int i;
for (i = 0; i < vec_upper_bound; i += 4)
  {
    int32x4_t vA = vld1q_s32 (&a[i]);
    int32x4_t vB = vld1q_s32 (&b[i]);
    vA += vB;
    vst1q_s32 (&res[i], vA);
  }
for (; i < n; i++)
  res[i] = a[i] + b[i];
```

Reply:
Ok I get it, this doesn't calculate the number of remaining items. It calculates what item the vector loop got up to.

So 6 >> 2 = 1, one loop processes 4 items. So when we get to the scalar we know 4 items have been processed. Then we count from that number up to the total count, then exit.

That's fine logic but it needs a comment, perhaps:
```
// The vector loop processes 4 items at a time, so there may be some remaining items.
// This sets w4 to the index of the next item to be processed.
lsl w4, w5, 2
``